azure-keyvault/src/main/java/com/microsoft/azure/keyvault/messagesecurity/HttpMessageSecurity.java

```diff
@@ -0,0 +1,330 @@
+     */
+    public Response unprotectResponse(Response response) throws IOException{
+        try{
+            if (!supportsProtection())
+                return response;
+
+            JWSObject jwsObject = JWSObject.deserialize(response.body().string());
+            JWSHeader jwsHeader = jwsObject.jwsHeader();
+
+            if(!jwsHeader.kid().equals(serverSignatureKey.kid()) || !jwsHeader.alg().equals("RS256")){
+                throw new IOException("Invalid protected response");
+            }
+
+            byte[] data = (jwsObject.original_protected() + "." + jwsObject.payload()).getBytes();
+            byte[] signature = MessageSecurityHelper.base64UrltoByteArray(jwsObject.signature());
+
+            RsaKey serverSignatureRsaKey = new RsaKey(serverSignatureKey.kid(), serverSignatureKey.toRSA(false));
+            boolean signed = serverSignatureRsaKey.verifyAsync(data, signature, "RS256").get();
+            if (!signed){
+                throw new IOException("Wrong signature.");
+            }
+
+            String decrypted = unprotectPayload(jwsObject.payload());
+
+            MediaType contentType = response.body().contentType();
+            ResponseBody body = ResponseBody.create(contentType, decrypted);
+            return response.newBuilder().body(body).build();
+        } catch (ExecutionException e){
+            // unexpected;
+            return null;
+        } catch (InterruptedException e){
+            // unexpected;
+            return null;
+        } catch (NoSuchAlgorithmException e){
+            // unexpected;
+            return null;
+        }
+    }
+
+    /**
+     * Return current timestamp. Returns always 0 for tests.
+     *
+     * @return
+     *      current timestamp or 0 for test mode.
+     */
+    private long getCurrentTimestamp(){
+        if (testMode){
+            return 0;
+        } else {
+            return System.currentTimeMillis() / 1000;
+        }
+    }
+
+    /**
+     * Check if HttmMessageSecurity has all required keys.
+     *
+     * @return
+     *      true if there is client signature key and two server keys.
+     */
+    private boolean supportsProtection(){
+        return this.clientSignatureKey != null &&
+               this.serverSignatureKey != null &&
+               this.serverEncryptionKey != null;
+    }
+
+
+    /**
+     * Encrypt provided payload and return proper JWEObject.
+     *
+     * @param payload
+     *      string to be encrypted.
+     *
+     * @return
+     *      JWEObject with encrypted payload.
+     */
+    private JWEObject protectPayload(String payload) throws IOException{
+        try{
+            JWEHeader jweHeader = new JWEHeader("RSA-OAEP", serverEncryptionKey.kid(), "A128CBC-HS256");
+
+            byte[] aesKeyBytes = new byte[32];
+            new Random().nextBytes(aesKeyBytes);
+            if (testMode){
+                aesKeyBytes = "TEST1234TEST1234TEST1234TEST1234".getBytes();
+            }
```

Comment:
I would move all the conditional testMode stuff out these methods. Perhaps encapsulate in private method getKeyBytes(int count) or something